source/load-balancers/load-balancers.rst

```diff
@@ -0,0 +1,535 @@
+:code:`ConnectionCheckOutStartedEvent`, and only 1 :code:`ConnectionCheckedOutEvent`
+or :code:`ConnectionCheckOutFailedEvent`. Similarly, the driver MUST only publish 1
+:code:`ConnectionCheckedInEvent`.
+
+Behaviour With Cursors
+^^^^^^^^^^^^^^^^^^^^^^
+
+When the driver is in load balancing mode and executing any cursor-initiating command, the driver
+MUST NOT check the connection back into the pool unless the command fails or the server
+returns a cursor ID of :code:`0` (i.e. all documents are returned in a single batch).
+Otherwise, the driver MUST continue to use the same connection for all subsequent
+:code:`getMore` commands for the cursor. The driver MUST check the connection back
+into the pool if the server returns a cursor ID of :code:`0` in a :code:`getMore`
+response (i.e. the cursor is drained). When the cursor's :code:`close` method is
+invoked, either explicitly or via an implicit resource cleanup mechanism, the driver
+MUST use the same connection to execute a :code:`killCursors` command if necessary
+and then check the connection back into the pool regardless of the result.
+
+If a :code:`getMore` fails with a network error, drivers MUST leave the connection pinned
+to the cursor. When the cursor's :code:`close` method is invoked, drivers MUST NOT execute
+a :code:`killCursors` command because the pinned connection is no longer valid and MUST
+return the connection back to the pool.
+
+Behaviour With Transactions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When executing a transaction in load balancing mode, drivers MUST follow the rules outlined
+in `Sharded Transactions <../transactions/transactions.rst#sharded-transactions>`__ with one
+exception: drivers MUST use the same connection for all commands in the transaction
+(excluding retries of commitTranscation and abortTransaction in some cases). Pinning
+to a single connection ensures that all commands in the transaction target the same
+service behind the load balancer. The rules for pinning to a connection and releasing
+a pinned connection are the same as those for server pinning in non-load balanced sharded
+transactions as described in `When to unpin <../transactions/transactions.rst#when-to-unpin>`__.
+
+Connection Tracking
+^^^^^^^^^^^^^^^^^^^
+
+The driver connection pool MUST track the purpose for which connections are checked out
+in the following 3 categories:
+
+- Connections checked out for cursors
+- Connections checked out for transactions
+- Connections checked out for operations not falling under the previous 2 categories
+
+When the connection pool's :code:`maxPoolSize` is reached and the pool times out waiting
+for a new connection the :code:`WaitQueueTimeoutError` MUST include a new detailed message,
+"Timeout waiting for connection from the connection pool. maxPoolSize: n, connections
+in use by cursors: n, connections in use by transactions: n, connections in use by other
+operations: n". A :code:`ConnectionCheckOutFailedEvent` MUST also be published with the
```

Comment:
Per the CMAP spec, `ConnectionCheckoutFailedEvent` events have `address` and `reason` fields. The `reason` is an enum and there's nothing about using custom strings. How do we expect drivers to include this new error message in the event as required by the last sentence of this paragraph?

Reply:
Oh I missed that part. What do you think about adding a new reason enum or reusing "poolClosed". I'm leaning towards the former here, maybe something like "poolClosedForService"?

Reply:
The spec uses "closed" to mean that the pool itself has been closed, which would probably be done via a call like `MongoClient.close()` in drivers. This doesn't fit with the scenario here, though, because the pool is still open. We just can't get a connection because we've reached the max size.

Also, aside from wording, I'd prefer to not add a new reason here. Reaching `maxPoolSize` is already possible today. When this happens, we return a `waitQueueTimeoutError` and publish an event with `reason=timeout`. I see no reason to change that - we're just making the error message more descriptive because it's "easier" to reach `maxPoolSize` in LB mode, but I don't think we need to change the event structure.

Reply:
Removed the reference to the event itself so behaviour remains the same.